### PR TITLE
BUG: regression in fortran array views due to relaxed strides

### LINKED ISSUE #6678: getting unexpected error from: X.dtype = complex (or Y = X.view(complex))
The following works with numpy 1.9.3 and python 3.4.3, but in numpy 1.10.1 and python 3.5.0 I get `ValueError: new type not compatible with array.` on the `b.dtype = complex` line:

    import numpy as np

    A = np.array([[1., 2.], [3., 4.]])
    A.dtype = complex          # this works

    a = np.array([[1., 2.]])
    a.dtype = complex          # this works

    B = np.array([[1., 2.], [3., 4.]], order='F')
    B.dtype = complex          # this works

    b = np.array([[1.], [2.]], order='F')
    b.dtype = complex   # <--  In 1.10.1, gives `ValueError: new type not compatible with array.`

    print('A = ', A)
    print('a = ', a)
    print('B = ', B)
    print('b = ', b)

The output for the older version is as it should be (I think):

    A =  [[ 1.+2.j]
     [ 3.+4.j]]
    a =  [[ 1.+2.j]]
    B =  [[ 1.+3.j  2.+4.j]]
    b =  [[ 1.+2.j]]

Note that the 1.10.1 gave the same results for `A`, `a`, and `B`.

My current workaround is to flatten `b`, set the dtype, and reshape. I’ve seen this same behavior on Linux and on Windows.

### PR DESCRIPTION
**Pull request by ahaldane**
This fixes #6678 so that views of fortran arrays with size-1 dims can be viewed as larger dtypes. It is backward compatible and keeps relaxed strides in place. 

At this point it is just to get an idea what it might look like. I am suggesting that for 1.10.2 we might use this PR but maybe with an extra deprecation warning (still to be decided) for anyone attempting such a view. Then in 1.11 we could implement the replacement functionality (eg, an 'order' keyword for views).

### CODE REVIEW

File: numpy/core/src/multiarray/getset.c [hunk 1]
```diff
@@ -419,6 +419,29 @@ array_nbytes_get(PyArrayObject *self)
...
+_use_C_view_dimension(PyArrayObject *ap)
+{
+    npy_intp sd;
+    npy_intp dim;
+    int i;
+    npy_bool is_c_contig = 1;
+
+    sd = PyArray_ITEMSIZE(ap);
+    for (i = PyArray_NDIM(ap) - 1; i >= 0; --i) {
+        dim = PyArray_DIMS(ap)[i];
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/getset.c]: There is a `PyArray_DIM` function that you can use to get a single dimension of an array, i.e. `PyArray_DIM(ap, i)` in this case.  That function does the exact same thing you are doing, but it is probably better to use our own API.

File: numpy/core/src/multiarray/getset.c [hunk 2]
```diff
@@ -419,6 +419,29 @@ array_nbytes_get(PyArrayObject *self)
...
+        if (PyArray_STRIDES(ap)[i] != sd) {
```
Comment (jaimefrio) [inline numpy/core/src/multiarray/getset.c]: Here you can similarly do `PyArray_STRIDE(ap, i)`.

**Comment by seberg** [comment]
This is weird logic, because `F` order is tried, even when the array is not in Fortran order? Is/was this intentional? Should we keep this? An "F if not C" logic (so the other way around) seems more reasonable to me?!

**Comment by ahaldane** [comment]
I cleaned it up with @jaimefrio's suggestion. 

@seberg I agree the logic is a little funny, but my goal here was to exactly reproduce the old behavior, to avoid shaking things up too much for 1.10.3.

**Comment by seberg** [comment]
The problem is that someone would check+ensure that the array is C-order before the `view` could in principle trigger Fortran order view like this. If the result dtype size is smaller than the original one, the result shape is different in that case I think.

**Comment by ahaldane** [comment]
@seberg I don't follow. If someone has ensured the array is C-order, I don't see how it's possible to trigger the fortran order view. The logic is "if C-ordered use last axis, otherwise use first axis", right?

**Comment by ahaldane** [comment]
Also, I was tempted to add the code below to the test. It checks that the new behavior matches the 1.9.3 behavior. 

I have tested on my system and it passes (suggesting I got the behavior right), BUT only after getting rid of the 'smokeout' dummy values in strides (like in #6342). The dummy values are screwing up my function `_use_C_view_dimension`, and I don't see any way to avoid this. 

So if I added the tests they would fail with current master, although (I am pretty sure) they would pass if backported to 1.10, which has removed the dummy values.


```python
# test arrays which are both C- and F- contiguous
f1x4 = np.array([[1,  2,  3,  4]], order='F', dtype='f4')
f4x1 = np.array([[1],[2],[3],[4]], order='F', dtype='f4')
c1x4 = np.array([[1,  2,  3,  4]], order='C', dtype='f4')
c4x1 = np.array([[1],[2],[3],[4]], order='C', dtype='f4')

assert_equal(f1x4.view('i2').shape, (2,4))
assert_equal(f4x1.view('i2').shape, (8,1))
assert_equal(c1x4.view('i2').shape, (1,8))
assert_equal(c4x1.view('i2').shape, (4,2))

assert_raises(f1x4.view('c8'), ValueError)
assert_equal( f4x1.view('c8').shape, (2,1))
assert_equal( c1x4.view('c8').shape, (1,2))
assert_raises(c4x1.view('c8'), ValueError)

# test arrays with many dimensions
a = np.empty((2,4,6), dtype='f4', order='C')
b = np.empty((2,4,6), dtype='f4', order='F')
c = np.empty((2,6,4), dtype='f4', order='C').transpose((0,2,1))
assert_equal(a.view('c8').shape, (2,4,3))
assert_equal(b.view('c8').shape, (1,4,6))
assert_raises(c.view('c8'), ValueError)
```

**Comment by seberg** [comment]
I think it was always a bit shaky maybe, but consider this:
```
arr = np.ones((1, 10), order='F', dtype=np.int64)
arr[0, :] = np.arange(10)

for_sure_c_order = np.asarray(arr, order='C')

view = arr.view(dtype=np.int32)
```

**Comment by ahaldane** [comment]
That's an interesting case I hadn't thought of! (And I assume you meant `for_sure_c_order` in the last line)

I'd interpret this as a quirk with `asarray`, rather than with `view`: If you examine the strides of `for_sure_c_order`, you will find that it is consistent with fortran ordering rather than c ordering (eg compare to the strides of `arr2 = np.ones((1, 10), order='C', dtype=np.int64)`). In other words `asarray` doesn't guarantee that the array will be "c-ordered", rather it only guarantees it will be "c-contiguous". 

This behavior hasn't changed between 1.9.3 and 1.10. So I think this PR is still totally consistent with 1.9.3.

**Comment by seberg** [comment]
Yes, meant for_sure_c_order there. And it *did* change.

Imagine you wrote a function which uses view and safeguard against non C-contiguous input (such as I did here) by the user. This may blow up currently and definitely will blow up with this. I don't care whether it is a quirk with asarray, it would mean also cleaning up all things that want C-order arrays to actually create a C-order view on them just to make `view` with different dtype happy.

**Comment by ahaldane** [comment]
Oh, you're right it did change, somehow I mis-tested it. And I agree it's a pretty big problem for this PR...

**Comment by seberg** [comment]
Its pretty scary logic (the old one) if anyone ever used fortran order.... And frankly I have no good idea right now how to fix it quite safe.

**Comment by ahaldane** [comment]
Looking again, in my opinion the behavior of `asarray` is wrong, or at least inconsistent with the doc. The docstring says:

```
order : {'C', 'F'}, optional
    Whether to use row-major (C-style) or
    column-major (Fortran-style) memory representation.
    Defaults to 'C'.
```

but that's not what it actually does any more. I think in 1.9.3 that was a correct description, but it no longer is in 1.10. I think (maybe?) if we change `asarray` to actually do what the doc says, then this PR is ok.

**Comment by ahaldane** [comment]
Just to add to that, I think the change in behavior of asarray should be considered a bug. Lets say someone is writing a C extension and uses the C version of `asarray` to order the array, and then iterate through the array manually with for loops based on the shape (ie, they assume C-strides). Then I think their code will be broken in 1.10, because `asarray` might not actually make it C-ordered.

**Comment by seberg** [comment]
@ahaldane the whole point of relaxed strides is to stop caring about strides that do not matter. The documentation is correct, the memory order of the data *is* C-contiguous, it is only the strides that look a bit funny when you are not careful (but you could always create things like 1x1 arrays, that broke).

Now you could clean up the strides and return a view to do just that, and it would be somewhat fine (we actually do this for the buffer protocol to some degree), but the `order` here and the `flags` also answer the question: "Can I use this as a C-array pointer", and the answer to this question is correct.

In other words, it is perfectly correct (and better, because you do not have unnecessary overhead!) for interfacing code, unless the interfacing code is weird (i.e. it says it uses C-order, but then uses a combination of strides and C-order that makes no sense).

**Comment by ahaldane** [comment]
I understand better, thanks.

If the strides do not matter, perhaps `asarray` should recompute the strides then? We could make it so asarray rearranges strides so that `for_sure_c_order` is still a view of `arr`, but has the same strides as my `arr2`.

**Comment by ahaldane** [comment]
All right, updated so `asarray` recomputes the strides if necessary. I think it is the right thing to do anyway, because I still think the first stride in `for_sure_c_order` didn't make sense. So at least as far as `view` and `asarray` are concerned I think/hope this PR now reproduces the 1.9.3 behavior!

Also, I added a commit to remove the dummy stride values so I can use my tests, we can remove it when it comes time.

**Comment by njsmith** [comment]
Nitpicks critiques: I'm pretty sure there are some refcount issues in the use/definition of recomputeStrides, because sometimes it returns a new object and sometimes it returns a borrowed reference. Also, we don't use camelCase :-)

Big picture critiques: I guess this approach might be OK as a temporary hack, but IMO we absolutely should not be assigning any special meaning to stride values beyond that `index @ strides = data offset`. So even if we decide to add hacks try to fix this for 1.10, we should be least be spitting out some deprecation warning and planning to remove the recomputation stuff entirely asap.

I have very mixed feelings about whether it's worth merging this. It still doesn't cover all the cases that 1.10 broke AFAICT. Like
```
a = np.ones((10, 1), order="F")
b = a + 1
```
IIUC, in 1.9 both a and b count as F ordered for view purposes, and this patch makes it so a will count as F ordered but does nothing for b?

**Comment by ahaldane** [comment]
I agree overall. Your example surprised me though, I guess I'm not aware of everything relaxed strides touches. Since you have a better idea than me, do you think it might eventually possible to reproduce the 1.9.3 behavior here by just a few further little tweaks like these? 

If we don't merge something like this PR, what are alternatives? To me it all seems scary since we would be breaking things without warning.

Also I am all ears on what to deprecate. I'd even be happy banning itemsize changes completely, since users can still use the `ndarray` method to reinterpret the memory however they want.

**Comment by seberg** [comment]
Yeah. relaxed strides cause "order" to be lost in more cases than before. And there is nothing you can do about it really. For interfacing code it could be dangerous if someone does not double check, but assumes contiguity stays (of course it is more complex there, since both `a` and `b` are fine unless you add broadcasting or things).
`view` is particularly nasty…. The `asarray` mitigation is ok with me, but I do not think there is a way to break no `view` usage except pulling out relaxed strides or spitting errors+adding `order` kwarg (or similar). Even deciding when to spit errors/warnings is non-trivial or not possible (but can probably be solved "good enough").

**Comment by seberg** [comment]
Frankly, I am lost on how to proceed with this. As I said, I currently see three options (in principle at least):
  1. Go with fixing up strides with asarray, and hope it is enough (could do that in any case). We could think about tweaking the view logic a bit also (not sure if that makes sense).
  2. Add the `order` keyword argument, and spit errors/warnings (accelerated deprecation)
  3. Revert relaxed strides and make `view` safe (order kwarg or such).

**Comment by ahaldane** [comment]
What do you all make of this:

    >>> (np.ones(2, order='F') + np.ones(2, order='F')[:,newaxis]).flags
      C_CONTIGUOUS : True
      F_CONTIGUOUS : False
      OWNDATA : True
      WRITEABLE : True
      ALIGNED : True
      UPDATEIFCOPY : False


With relaxed strides adding two fortran contiguous arrays gives a c-ordered array.

**Comment by ahaldane** [comment]
After sitting on that a little, here is what I'm thinking:

Relaxed strides was a step in the right direction - its new idea was that by looking at *contiguity* rather than *order* we could avoid a lot of costly memory copies and buffering. But maybe it went a step too far by removing *order*, replacing it completely with *contiguity*. 

There are two known cases now where you do need to know order: In certain views, and when determining the output order of ufuncs. The problem arises in general for operations which increase the number of dimensions, when the inputs are both C- and F- contiguous (ie, arrays with only 1 big dim), and therefore have unknown order. This problem probably also arises for any other operations which can increase the number of dimensions.

I think it may be possible to build on the relaxed strides changes by keeping track of both *order* and *contiguity*, giving us both the performance enhancements of relaxed strides by knowing contiguity, plus consistency from knowing order. It would probably look something like this PR, actually.

**Comment by charris** [comment]
@ahaldane ufuncs have and order keyword
```
In [3]: np.add(eye(2), eye(2), order='F').flags
Out[3]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False
```

**Comment by ahaldane** [comment]
Hmm OK. Well another possibility I was thinking about is having no guarantee on the output order of ufuncs, and that's made more sensible if you can force the order with a keyword.

In that case, if we make no guarantees about the output order of `a+b`, then there is no problem with @njsmith's example above, and I don't need touch ufuncs in this PR.

The only thing that worries me about that is that there does seem to have been some effort to guarantee the output order in operations like `a+b` in 1.9.3. I was just playing around with arrays in various weird orders (not even C- or F- ordered) and 1.9.3 does a good job preserving the order in the output. Eg,

    >>> a = np.random.rand(2,2,2).transpose((0,2,1))
    >>> a.strides, (a+a).strides
    ((32, 8, 16), (32, 8, 16))

**Comment by charris** [comment]
I don't know what the default order was before, suspect 'C'. There is also 'K', which could be a better choice but might break things.
```
In [2]: np.add(eye(2).T, eye(2).T, order='K').flags
Out[2]: 
  C_CONTIGUOUS : False
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  UPDATEIFCOPY : False
```

**Comment by seberg** [comment]
Before 1.7, I am not sure there was K really, now the default is K, I am pretty sure.

**Comment by seberg** [comment]
Well.... I have always been on the "no guarantees" side when it comes to the order from strides. For `view` it is particularly difficult, and relaxed strides might create more problems in some cases then even 1.6 to 1.7, because it affects "clean" Fortran arrays in some cases (but the only complains I remember from 1.7 were recently by josef I think).
That said, on the other hand relying on these things was never quite save in all cases as well.

Anyway, I find that I have still no idea how to best proceed, nor right now the strength to fight for relaxed strides in this release.

So, we have option 1. We keep relaxed strides, mitigate some issues in `view` (basically this PR as it is), and possibly deprecate or put warnings on the rest and hope for the best. Which also means hope for the best with possible output order changes that seemed to have startled even some of us.
The other option, is undoing the relaxed strides for better or worse (the one real advantage is, that my brain can stop thinking about it ^^).

In any case, I think we have to
  1. Send a mail to the list, so that some more people get a chance to protest (in case 1) or comment in general.
  2. If we go with reverting for now (for whatever reason), I would propose that we actually write an NEP to put the stuff somewhere more formally.
Not that I have much hope of getting serious comments.

If anyone thinks it might help, we could also do a conference talk, but to me, the only thing that would really help is hearing from many people that they don't mind relaxed strides, or from one that it would utterly break their stuff ;).

**Comment by charris** [comment]
My inclination is to leave as is, which will cause a few folks inconvenience but a least raise an error, and add the `order` keyword to view. For backward compatibility we can recommend the transpose-view-tranpose route. Doing that isn't strictly legal, but I think it gets us out of this jam while letting us keep relaxed strides. The other option is to make relaxed strides optional and start with a deprecation warning and I'd as soon avoid that if possible.

**Comment by charris** [comment]
Note that the only problems for 1.10.2 are arrays that are both C and F ordered and we can raise a warning for those cases that the default view is C ordered, and for F order only we might want to raise a FutureWarning.

@njsmith @rgommers Thoughts?

**Comment by seberg** [comment]
Sounds OK, to me. So that would mean we add an `order='warn'` (or None) kwarg, that we will switch to `order='C'` as default?

**Comment by charris** [comment]
Just to analyze this a bit, in 1.9.x the complex view works only if the first dimension is a multiple of two and the array is `F_CONTIGUOUS` *or* the last dimension is a multiple of two and the array is  `C_CONTIGUOUS`. Note that  `F_CONTIGUOUS` and `C_CONTIGUOUS` are mutually exclusive except for 1-D arrays or arrays with all dimensions equal to 1.
```
Out[1]: array([[ 1.+1.j]])

In [2]: ones((2,1)).view(complex)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-f64f3d9e2c62> in <module>()
----> 1 ones((2,1)).view(complex)

ValueError: new type not compatible with array.

In [3]: ones((1,2), order='F').view(complex)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-46c34c42a4c2> in <module>()
----> 1 ones((1,2), order='F').view(complex)

ValueError: new type not compatible with array.

In [4]: ones((2,1), order='F').view(complex)
Out[4]: array([[ 1.+1.j]])

In [5]: ones((2,)).view(complex)
Out[5]: array([ 1.+1.j])

In [6]: ones((2,), order='F').view(complex)
Out[6]: array([ 1.+1.j])

In [7]: ones((1,1)).view(int8).shape
Out[7]: (1, 8)

In [8]: ones((1,1), order='F').view(int8).shape
Out[8]: (1, 8)
```
Consequently, the problem with relaxed strides only happens when an array is both `C_CONTIGUOUS` and `F_CONTIGUOUS` and is of dimension > 1, which can only happen when at most a single dimension is > 1. Furthermore, if that index is the last index or all dimensions are 1, we are backward compatible. So the problem is only if the first index is > 1 and the array has > 1 dimension. I think we can special case that and be good to go, except that we will actually allow *more* arrays to be treated a Fortran arrays without taking the trouble to explicitly mark them so.

If we take that approach, then for 1.11 we could deprecate that behavior and add an order keyword to view. I'd also suggest deprecating assignment to `a.dtype`. Note that the 1.9.3 behavior above for `ones((1,1), order='F').view(int8).shape` looks wrong as is.



For only problem we need to deal with for relaxed strides is when the array is both `F_CONTIGUOUS` and `C_CONTIGUOUS`. That case only happens when at most a single dimension is > 1.

**Comment by seberg** [comment]
Yes, but I am not sure you covered the cases of a *smaller* new dtype, which expands the dimensions differently in the limited cases you listed, but is a bit worse:
```
# On 1.8:
In [8]: np.ones((1, 2), dtype=np.int64).view(dtype=np.int32)
Out[8]: array([[1, 0, 1, 0]], dtype=int32)

In [9]: np.ones((1, 2), dtype=np.int64, order='F').view(dtype=np.int32)
Out[9]: 
array([[1, 1],
       [0, 0]], dtype=int32)

# On master:
In [1]: np.ones((1, 2), dtype=np.int64).view(dtype=np.int32)
Out[1]: array([[1, 0, 1, 0]], dtype=int32)

In [2]: np.ones((1, 2), dtype=np.int64, order='F').view(dtype=np.int32)
Out[2]: array([[1, 0, 1, 0]], dtype=int32)
```

**Comment by charris** [comment]
Urg. OK, there is simply no way to get that, even from strides. The problem, as @ahaldane has said, it that the original flags were used to signal intent as well as contiguity, i.e., this array is f_contiguous *and* treat it as a fortran array. From that point of view the 1.9 behavior
```
In [8]: ones((1,1), order='F').view(int8).shape
Out[8]: (1, 8)
```
definitely looks wrong ;) So if we stick with relaxed stride we either need to add a flag (not backward compatible), or add a view keyword (not backward compatible), or raise an warning in the ambiguous cases and ask folks to use the `a.T.view(...).T` workaround if they really, really, need the old behavior. Or some combination of those.

**Comment by charris** [comment]
I'm tending towards not making relaxed stride checking the default for 1.10. I've added an `order` keyword to the view method to see how things look, but since `PyArray_View` is in the API, we need another version of that also. That, plus arrays with shape `(1,1,1,1,...)` behaving differently with `order='F'` makes me think that more time is needed to prepare for the change.

**Comment by seberg** [comment]
I am completely fine with that, it is the only straight forward solution without "hard to estimate how dangerous" traps.

**Comment by charris** [comment]
OK, so for 1.10.x I will change the default. Meanwhile, @ahaldane I think it would be good to separate out the new tests and get them into master. A boolean like
```
is_relaxed_strides = np.ones((2, 1), order='C').flags.f_contiguous
```
might be useful there. Could probably be put into the `numpy/__init__.py` file so everyone could use it.

**Comment by charris** [comment]
@ahaldane @seberg I'm closing this. Please reopen if you want to keep it active.
